pyinterprod/interpro/signature.py

```diff
@@ -546,3 +574,120 @@ def update_features(uri: str, update: list[tuple[Database, list[str]]]):
     con.commit()
     cur.close()
     con.close()
+
+
+def get_pmid2pubid(cur: cx_Oracle.Cursor) -> dict[int, str]:
+    cur.execute(
+        """
+        SELECT DISTINCT PUBMED_ID, PUB_ID 
+        FROM INTERPRO.CITATION
+        """
+    )
+    return dict(cur.fetchall())
+
+
+def get_method2pub(cur: cx_Oracle.Cursor) -> dict[str, set]:
+    cur.execute(
+        """
+        SELECT METHOD_AC, LISTAGG(PUB_ID, ';')
+        FROM INTERPRO.METHOD2PUB
+        GROUP BY METHOD_AC
+        """
+    )
+
+    current_method2pub = {}
+    for method_ac, pub_ids in cur:
+        current_method2pub[method_ac] = set(pub_ids.split(";"))
+    return current_method2pub
+
+
+def update_references(cur: cx_Oracle.Cursor, method: Method,
+                      pmid2pubid: dict[int, str], new_method2pub: dict):
+    if method.abstract is not None:
+        pmids = re.findall(r"PMID:\s*([0-9]+)", method.abstract)
+        for pmid in map(int, pmids):
+            try:
+                pub_id = pmid2pubid[pmid]
+            except KeyError:
+                pub_id = update_citation(cur, pmid)
+                pmid2pubid.update({pmid: pub_id})
+
+            if pub_id:
+                method.abstract = method.abstract.replace(f'PMID:{pmid}', f'[cite:{pub_id}]')
+                new_method2pub[method.accession].add(pub_id)
+
+    for pmid in method.references:
+        try:
+            pub_id = pmid2pubid[pmid]
+        except KeyError:
+            pub_id = update_citation(cur, pmid)
+            pmid2pubid.update({pmid: pub_id})
+
+        if pub_id:
+            new_method2pub[method.accession].add(pub_id)
+
+
+def update_method2pub(cur: cx_Oracle.Cursor, method2pub: dict[str, set]):
+    method_ids = [i for sublist in [[(m, pmid) for pmid in pmdis] for m, pmdis in method2pub.items()] for i in sublist]
+    cur.executemany(
+        """
+        INSERT INTO INTERPRO.METHOD2PUB_STG (PUB_ID, METHOD_AC)
+        VALUES (:1, :2)
+        """,
+        method_ids
+    )
```

Comment:
Oracle may raise an error if there are _too many_ rows being inserted at the same time (not sure what is the limit), so we can insert rows by chunks of 1k:

```suggestion
    step = 1000
    for i in range(0, len(data), step):
        cur.executemany(
            """
            INSERT INTO INTERPRO.METHOD2PUB_STG (PUB_ID, METHOD_AC)
            VALUES (:1, :2)
            """,
            data[i:i+step]
        )
```